src/mlpack/methods/svdplusplus/svdplusplus_function_impl.hpp

```diff
@@ -0,0 +1,561 @@
+  //     rank - 1, numUsers + 2 * numItems - 1)
+  // Unused:
+  //     row(rank).subvec(numUsers + numItems, numUsers + 2 * numItems - 1)
+  initialPoint.randu(rank + 1, numUsers + 2 * numItems);
+}
+
+template<typename MatType>
+void SVDPlusPlusFunction<MatType>::Shuffle()
+{
+  data = data.cols(arma::shuffle(arma::linspace<arma::uvec>(0, data.n_cols - 1,
+      data.n_cols)));
+}
+
+template <typename MatType>
+double SVDPlusPlusFunction<MatType>::Evaluate(const arma::mat& parameters) const
+{
+  return Evaluate(parameters, 0, data.n_cols);
+}
+
+template <typename MatType>
+double SVDPlusPlusFunction<MatType>::Evaluate(const arma::mat& parameters,
+                                              const size_t start,
+                                              const size_t batchSize) const
+{
+  // The cost for the optimization is as follows:
+  //    f(u, v, p, q, y) =
+  //        sum((rating(i, j) - p(i) - q(j) - u(i).t() * (v(j) + sum(y(k))))^2)
+  // The sum is over all the ratings in the rating matrix.
+  // 'i' points to the user and 'j' points to the item being considered.
+  // 'k' points to the items which user 'i' interacted with.
+  // The regularization term is added to the above cost, where the vectors u(i)
+  // and v(j), bias p(i) and q(j), implicit vectors y(k) are regularized for
+  // each rating they contribute to.
+
+  // It's possible this loop could be changed so that it's SIMD-vectorized.
+  double objective = 0.0;
+
+  // The norm square of implicit item vectors is cached to avoid repeated
+  // calculation.
+  arma::vec implicitVecsNormSquare(numItems);
+  implicitVecsNormSquare.fill(-1);
+
+  for (size_t i = start; i < start + batchSize; ++i)
+  {
+    // Indices for accessing the the correct parameter columns.
+    const size_t user = data(0, i);
+    const size_t item = data(1, i) + numUsers;
+    const size_t itemRealIdx = data(1, i);
+    const size_t implicitStart = numUsers + numItems;
+
+    // Calculate the squared error in the prediction.
+    const double rating = data(2, i);
+    const double userBias = parameters(rank, user);
+    const double itemBias = parameters(rank, item);
+
+    // Iterate through each item which the user interacted with to calculate
+    // user vector.
+    arma::vec userVec(rank, arma::fill::zeros);
+    arma::sp_mat::const_iterator it = implicitData.begin_col(user);
+    arma::sp_mat::const_iterator it_end = implicitData.end_col(user);
+    size_t implicitCount = 0;
+    double regularizationError = 0;
+    for (; it != it_end; it++)
+    {
+      userVec += parameters.col(implicitStart + it.row()).subvec(0, rank - 1);
+      if (implicitVecsNormSquare(itemRealIdx) < 0)
+      { 
+        implicitVecsNormSquare(itemRealIdx) = arma::dot(
+            parameters.col(implicitStart + it.row()).subvec(0, rank - 1),
+            parameters.col(implicitStart + it.row()).subvec(0, rank - 1));
+      }
+      regularizationError += lambda * implicitVecsNormSquare(itemRealIdx);
```

Comment:
I think, the code above is incorrect (you should use `it.row()` instead of `itemRealIdx`). The fix follows:
```
if (implicitVecsNormSquare(it.row()) < 0)
{
  implicitVecsNormSquare(it.row()) = arma::dot(
      parameters.col(implicitStart + it.row()).subvec(0, rank - 1),
      parameters.col(implicitStart + it.row()).subvec(0, rank - 1));
}
regularizationError += lambda * implicitVecsNormSquare(it.row());
```

Reply:
Yes you are right. I will fix it. Thanks!